### PR TITLE
feat: add automatic AB# tagging from branch name

### LINKED ISSUE #151: Automatically add AB# tag from id in source branch
Loving the latest enhancements by the way, thanks!

I have another feature request that would really help automation and reduce developer toil.
Can you add a new option that attempts to automatically add the AB# link to the PR body/title by scanning the head branch name for a work item id

e.g.
I create a PR to merge branch `task/12345/make-it-better` into main
`AB#12345` is added to the PR body

I guess there will be other variations of branch name that contain the work item id
 `task/12345/make-it-better` 
 `task/12345-make-it-better` 
 `task/12345` 
 `task-12345` 
 `12345-make-it-better` 
 `12345make-it-better` 
 `12345`

### PR DESCRIPTION
**Pull request by joshjohanning**
## Summary

Add new `add-work-item-from-branch` feature that extracts work item IDs from the head branch name and adds `AB#xxx` to the PR body. Includes three layers of false-positive protection and always validates IDs against Azure DevOps before adding.

Resolves #151

## New Inputs

| Input | Default | Description |
|---|---|---|
| `add-work-item-from-branch` | `false` | Enable automatic work item extraction from branch names |
| `branch-work-item-prefixes` | `task, bug, bugfix` | Comma-separated keyword prefixes that identify work item IDs |
| `branch-work-item-min-digits` | `5` | Minimum digit length for extracted IDs (set to `5` if your IDs are 10000+) |

## False Positive Protection

Three safeguards prevent incorrect linking:

1. **Keyword prefix matching** — only numbers following a configured keyword (`task`, `bug`, `bugfix`) and a separator (`/`, `-`, `_`) are extracted
2. **First-segment-only** — keywords must appear at the start of the branch or after a path separator (`/`). Keywords in description segments are ignored (e.g. `task/12345/fix-bug-67890` → only `12345`, not `67890`)
3. **Date pattern exclusion** — numbers followed by date-like patterns (e.g. `2024-01-15`) are automatically rejected
4. **Minimum digit length** — configurable via `branch-work-item-min-digits` to filter short numbers (years, versions)
5. **Azure DevOps validation** — every extracted ID is validated against Azure DevOps before being added; invalid IDs are skipped with a warning

## Branch Name Examples

| Branch Name | Extracted IDs | Notes |
|---|---|---|
| `task/12345/fix-login` | `12345` | ✅ Standard convention |
| `bug-67890` | `67890` | ✅ Keyword with `-` separator |
| `users/josh/task/12345/fix` | `12345` | ✅ Keyword after path separator |
| `bug/12345/task/67890-combined` | `12345`, `67890` | ✅ Multiple keywords as path segments |
| `bugfix/2024-01-15-fix-login` | _(none)_ | ⛔ Date pattern excluded |
| `task/12345/fix-bug-67890` | `12345` | ⛔ `bug` in description ignored |
| `hotfix/2024-bugfix` | _(none)_ | ⛔ `hotfix` is not a default prefix |
| `feature/task-manager-ui` | _(none)_ | ⛔ No number after keyword |
| `task-12345-bug-67890` | `12345` | ⛔ `bug` after `-` ignored (not a path prefix) |

## Additional Fixes

- **`validateWorkItemExists` return type** — branch-tagging path now correctly handles `{ exists, authError?, errorMessage? }` object return instead of treating it as a boolean
- **`AB_PATTERN.lastIndex` bug** — reset `lastIndex` before `.test()` calls on the module-level `/g` regex to prevent false negatives when validating multiple commits in a loop
- **`escapeRegExp` deduplication** — hoisted the function and reused it in `buildBranchWorkItemPattern` instead of duplicating the regex escape logic
- **Empty prefixes validation** — action now fails fast with a clear error when `add-work-item-from-branch` is enabled but `branch-work-item-prefixes` is empty
- **Standalone operation** — `add-work-item-from-branch` can run without `check-commits` or `check-pull-request` enabled
- **Sanitized branch name in job summary** — backticks and backslashes in branch names are escaped to prevent markdown injection
- **README** — added dedicated "Auto-Tag from Branch Name" section with examples and configuration guidance

## Tests

122 tests pass (29 bash + 93 Jest), including:
- 23 unit tests for `extractWorkItemIdsFromBranch` covering keyword matching, date exclusion, first-segment-only, min-digits, case-insensitivity, custom prefixes, and edge cases
- 12 integration tests for the full `add-work-item-from-branch` flow including auth error handling

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -701,6 +711,108 @@
...
+
+  core.info(`Adding work item tag(s) to PR body: ${abTags}`);
+  await octokit.rest.pulls.update({
+    owner,
+    repo,
+    pull_number: pullNumber,
+    body: updatedBody
+  });
+  core.info('PR body updated with work item tag(s) from branch name');
+  const sanitizedBranchName = branchName.replace(/`/g, '\\`');
```
[the remark this answers is not in the document]
Comment (joshjohanning) [reply src/index.js]: Fixed in 13c0a3a — `sanitizeBranchName()` now escapes both backticks and backslashes before interpolating into the job summary markdown.

File: src/index.js [hunk 2]
```diff
@@ -89,6 +110,18 @@ export async function run() {
 
     const octokit = github.getOctokit(githubToken);
 
+    // Automatically add AB# tags from branch name if enabled
+    if (addWorkItemFromBranch) {
+      await addWorkItemsToPRBody(
+        octokit,
+        context,
```
[the remark this answers is not in the document]
Comment (joshjohanning) [reply src/index.js]: Acknowledged — this is a single extra API call per run and doesn't materially affect latency or rate-limit pressure. Deferring this optimization as low priority.

File: action.yml [hunk 3]
```diff
@@ -48,6 +48,18 @@ inputs:
...
+    required: false
+    default: 'false'
+  branch-work-item-prefixes:
+    description: 'Comma-separated list of keyword prefixes used to identify work item IDs in branch names (e.g. task/12345). Only numbers following one of these keywords (separated by /, -, or _) are extracted. Only used when add-work-item-from-branch is true.'
+    required: false
+    default: 'task, bug, bugfix'
+  branch-work-item-min-digits:
+    description: 'Minimum number of digits for a work item ID extracted from a branch name. Set to 1 to match any length. Only used when add-work-item-from-branch is true.'
+    required: false
+    default: '5'
```
[the remark this answers is not in the document]
Comment (joshjohanning) [reply action.yml]: updated PR
